Add support for sparse jacobians in the scipy-interface

jhelgert opened the pull request:
Closes #98.

This PR adds support for sparse jacobians in the scipy interface.

- The user can pass dense and sparse constraint jacobians and even mix both
- Each provided jacobian is expected to be a function that returns either a `np.ndarray`(dense) or a `scipy.sparse.coo_array` (sparse).
- Under the hood, all jacobians are converted to sparse matrices in the COO format, i.e. [`coo_array`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.coo_array.html#scipy.sparse.coo_array).

PS: After this PR, the three functions `get_constraint_bounds`, `get_constraint_dimensions`, and `get_sparse_jacobian_structure` will be called in order to create the Ipopt NLP problem before the problem is passed to Ipopt. Since each of these functions will evaluate the constraints and jacobians, I think adding a note in the docs makes sense, specifically that the Problem-interface will have better performance for time-critical use cases or for large-scale problems where evaluating the functions is really expensive.

moorepants commented on cyipopt/scipy_interface.py:
```diff
@@ -185,12 +216,46 @@ def get_bounds(bounds):
         return lb, ub
 
 
+def get_sparse_jacobian_structure(constraints, x0):
```
This would be considered a public function. Either add a full docstring and include in the API docs or preprend the name with an underscore to make it private.

jhelgert replied:
Done.

moorepants commented on cyipopt/scipy_interface.py:
```diff
@@ -185,12 +216,46 @@ def get_bounds(bounds):
         constraints = (constraints, )
     for con in constraints:
-        if con.get('jac', False) is True:
+        if con.get('jac', False) == True:
```
```suggestion
        if con.get('jac', False):
```

jhelgert replied:
@moorepants Your proposed change is not equivalent to `if con.get('jac', False) == True`. If `con['jac']` is a function or a lambda, `if con.get('jac', False):` will be unintentionally true. However, it should be true if and only if `con['jac']` is a `bool` and `True`. So I'd suggest reversing it again.

PS: I noticed that some of the tests are (correctly) failing on my local machine after this commit. Any idea why this isn't the case for the GitHub workflows? I double-checked it and the tests should definitely fail...

brocksam replied:
The booleans in Python are built-in constants. Using an identity check (`is`) with `True` or `False` is idiomatic (it's more explicit and it's faster) as it checks object IDs rather than calling an object's `__eq__` method, which is what happens during an equality check (`==`). As you've said it should only be true iff `con['jac']` is `True` then it should be reverted to its state prior to this PR of `if con.get('jac', False) is True:`.

moorepants replied:
Sorry for leading you astray.

jhelgert commented:
@moorepants Thanks for the review. Does this PR look good to you or is there something left to work on before it can be merged?

moorepants commented:
Are there any performance negatives to converting the dense matrices to the sparse objects, even if they aren't sparse?

moorepants commented on cyipopt/scipy_interface.py:
```diff
@@ -76,7 +88,10 @@ def __init__(self,
+                 con_dims=(),
+                 sparse_jacs=(),
+                 jac_nnz_row=(),
+                 jac_nnz_col=()):
```
These new kwargs need to be documented.

moorepants commented:
It would be nice if an example is added to the documentation (at the minimum in the docstring here).

Other than that and the above comments, I think this is fine as long as there is no large computation cost to converting dense matrices to sparse ones.

brocksam commented on cyipopt/scipy_interface.py:
```diff
@@ -154,11 +171,26 @@ def constraints(self, x):
         return np.hstack(con_values)
 
+    def jacobianstructure(self):
+        return self._constraint_jacobian_structure
```
In the regular `Problem` interface, `jacobianstructure` and `hessianstructure` are callable methods, not properties. It's worth maintaining consistency with that.

brocksam commented:
> It would be nice if an example is added to the documentation (at the minimum in the docstring here).

Agreed. Inclusion of an example and a short explanation of this in the documentation would be a very valuable addition to this PR.

> Other than that and the above comments, I think this is fine as long as there is no large computation cost to converting dense matrices to sparse ones.

While there will be a performance hit in dense-to-sparse conversions, I'm almost certain Ipopt does this internally anyway as it can only deal with sparse Jacobians and Hessians. Also, of the sparse matrix formats COO is an efficient one to construct so the performance penalty here shouldn't be too large.

Ipopt shines on large sparse NLPs so I'd expect most users aren't working with dense problems. If a user is solving a small dense NLP then the conversion-in-Python vs conversion-in-C++ performance is likely not an issue. If a user is solving a large dense NLP then they should likely be using a different NLP solver.

jhelgert commented:
> Are there any performance negatives to converting the dense matrices to the sparse objects, even if they aren't sparse?

As Sam mentioned, Ipopt already converts all (dense) jacobians to sparse ones. Once the jacobian is evaluated from the Python interface, there shouldn't be a difference. However, there's definitely a performance penalty on the python side for problems containing *only dense* jacobians (or *many jacobians*) due to the dense-to-sparse conversions.

I ran some benchmarks that measure the time required to evaluate IpoptProblemWrapper's `jacobian` method. I used two dense jacobians, one with shape `5000 x N` and the other one with shape `5000 x N/2`. (You can find the code [here](https://gist.github.com/jhelgert/3c48176005a96d1023e6d41a13fe2dd5)):

![perf](https://user-images.githubusercontent.com/8471326/208711371-09bcc148-96f1-4d97-8df5-e6ad844a96b4.png)


- old: That's the status quo before the PR. Only dense jacobians are accepted and a large dense jacobian is passed to Ipopt.
- new1: The initial status of this PR (before the last commit that improved the performance). All dense/sparse jacobians are converted to a large sparse jacobian.
- new2: The status quo of this PR. The last commit improved the performance significantly.

IMO, the performance penalty is negligible. And I fully agree with Sam. @moorepants What do you think?

moorepants commented:
> IMO, the performance penalty is negligible. And I fully agree with Sam. @moorepants What do you think?

Thanks for the careful check. I'm fine with this penalty.

jhelgert commented:
@moorepants I added an example to the documentation which demonstrates how to use the sparse jacobians and simplified the `jacobian` method (new3). More than welcome side effect: No more performance penalty 🎉.

![perf](https://user-images.githubusercontent.com/8471326/208978899-634133e2-6b03-4743-a657-787a3fc25538.png)

jhelgert commented:
@moorepants Do you think it can be merged at the current state or is there something left to do on my part?

brocksam commented:
Sorry for the delay, @jhelgert. Like others, I had some good time away from the laptop over the festive period. Thank you for adding the comprehensive example to `tutorial.rst`, looks good. I'll give @moorepants another day or so to chime in with anything else, if there's nothing then I'll merge this.

moorepants commented:
LGTM
